File: account/src/main/java/serp/project/account/kernel/utils/JwtUtils.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.kernel.utils;

import com.nimbusds.jose.crypto.RSASSAVerifier;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import serp.project.account.kernel.property.KeycloakProperties;

import java.security.interfaces.RSAPublicKey;
import java.util.*;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
@Slf4j
public class JwtUtils {
    private final KeycloakJwksUtils keycloakJwksUtils;
    private final KeycloakProperties keycloakProperties;

    @SuppressWarnings("unchecked")
    public List<String> getRolesFromToken(String token) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);
            List<String> roles = new ArrayList<>();

            Object realmAccess = claimsSet.getClaim("realm_access");
            if (realmAccess instanceof Map) {
                Map<String, Object> realmAccessMap = (Map<String, Object>) realmAccess;
                Object realmRoles = realmAccessMap.get("roles");
                if (realmRoles instanceof List) {
                    roles.addAll((List<String>) realmRoles);
                }
            }

            Object resourceAccess = claimsSet.getClaim("resource_access");
            if (resourceAccess instanceof Map) {
                Map<String, Object> resourceAccessMap = (Map<String, Object>) resourceAccess;
                for (Object clientAccess : resourceAccessMap.values()) {
                    if (clientAccess instanceof Map) {
                        Map<String, Object> clientAccessMap = (Map<String, Object>) clientAccess;
                        Object clientRoles = clientAccessMap.get("roles");
                        if (clientRoles instanceof List) {
                            roles.addAll((List<String>) clientRoles);
                        }
                    }
                }
            }

            return roles.stream()
                    .filter(Objects::nonNull)
                    .distinct()
                    .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Error extracting roles from token", e);
            return Collections.emptyList();
        }
    }

    public boolean hasRole(String token, String roleName) {
        List<String> roles = getRolesFromToken(token);
        return roles.contains(roleName) || roles.contains(roleName.toUpperCase());
    }

    public Long getUserIdFromToken(String token) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);

            Object userIdClaim = claimsSet.getClaim("uid");
            if (userIdClaim instanceof Number) {
                return ((Number) userIdClaim).longValue();
            }
            if (userIdClaim instanceof String) {
                try {
                    return Long.parseLong((String) userIdClaim);
                } catch (NumberFormatException ignored) {
                    log.warn("uid claim is not a valid number: {}", userIdClaim);
                }
            }
            return null;
        } catch (Exception e) {
            log.error("Error extracting user ID from token", e);
            return null;
        }
    }

    public String getSubjectFromToken(String token) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);
            return claimsSet.getSubject();
        } catch (Exception e) {
            log.error("Error extracting subject from token", e);
            return null;
        }
    }

    public boolean isBearerToken(String token) {
        try {
            String typ = (String) extractClaim(token, "typ");
            return "Bearer".equals(typ);
        } catch (Exception e) {
            log.error("Error extracting token type from token", e);
            return false;
        }
    }

    public JWTClaimsSet validateToken(String token) {
        try {
            SignedJWT signedJWT = SignedJWT.parse(token);
            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();

            if (claimsSet.getExpirationTime() != null && claimsSet.getExpirationTime().before(new Date())) {
                throw new RuntimeException("JWT token has expired");
            }

            if (claimsSet.getIssueTime() != null && claimsSet.getIssueTime().after(new Date())) {
                throw new RuntimeException("JWT token not yet valid");
            }

            String issuer = claimsSet.getIssuer();
            if (issuer == null || issuer.trim().isEmpty()) {
                throw new RuntimeException("JWT token missing issuer");
            }

            if (keycloakProperties.getExpectedIssuer() != null &&
                    !keycloakProperties.getExpectedIssuer().equals(issuer)) {
                throw new RuntimeException("JWT token issuer mismatch. Expected: " +
                        keycloakProperties.getExpectedIssuer() + ", Actual: " + issuer);
            }

            List<String> audiences = claimsSet.getAudience();
            if (audiences == null || audiences.isEmpty()) {
                log.warn("JWT token has no audience claim");
            } else if (keycloakProperties.getExpectedAudience() != null &&
                    !audiences.contains(keycloakProperties.getExpectedAudience())) {
                throw new RuntimeException("JWT token audience mismatch. Expected: " +
                        keycloakProperties.getExpectedAudience() + ", Actual: " + audiences);
            }

            return validateKeycloakToken(signedJWT, claimsSet);

        } catch (Exception e) {
            log.error("Error validating JWT token: {}", e.getMessage());
            throw new RuntimeException("Invalid JWT token", e);
        }
    }

    private JWTClaimsSet validateKeycloakToken(SignedJWT signedJWT, JWTClaimsSet claimsSet) {
        try {
            String keyId = signedJWT.getHeader().getKeyID();
            if (keyId == null) {
                log.warn("No key ID found in JWT header, skipping signature verification");
                return claimsSet;
            }

            RSAPublicKey keycloakPublicKey = keycloakJwksUtils.getPublicKey(keyId);
            if (keycloakPublicKey == null) {
                log.warn("Could not find public key for key ID: {}, skipping signature verification", keyId);
                return claimsSet;
            }

            RSASSAVerifier verifier = new RSASSAVerifier(keycloakPublicKey);
            if (!signedJWT.verify(verifier)) {
                throw new RuntimeException("Invalid Keycloak JWT signature");
            }

            log.debug("Successfully verified Keycloak token signature with key ID: {}", keyId);
            return claimsSet;

        } catch (Exception e) {
            log.error("Error validating Keycloak token", e);
            throw new RuntimeException("Invalid Keycloak token", e);
        }
    }

    public boolean isTokenExpired(String token) {
        try {
            SignedJWT signedJWT = SignedJWT.parse(token);
            Date expirationDate = signedJWT.getJWTClaimsSet().getExpirationTime();
            return expirationDate != null && expirationDate.before(new Date());
        } catch (Exception e) {
            log.error("Error checking token expiration", e);
            return true;
        }
    }

    public JWTClaimsSet getClaimsFromToken(String token) {
        try {
            SignedJWT signedJWT = SignedJWT.parse(token);
            return signedJWT.getJWTClaimsSet();
        } catch (Exception e) {
            log.error("Error parsing claims from token", e);
            return null;
        }
    }

    public Object extractClaim(String token, String claimName) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);
            return claimsSet.getClaim(claimName);
        } catch (Exception e) {
            log.debug("Failed to validate token, trying without signature validation", e);
            // Fallback to extraction without signature validation
            return extractClaimWithoutValidation(token, claimName);
        }
    }

    public String getEmailFromToken(String token) {
        try {
            return (String) extractClaim(token, "email");
        } catch (Exception e) {
            log.error("Error extracting email from token", e);
            return null;
        }
    }

    public String getPreferredUsernameFromToken(String token) {
        try {
            return (String) extractClaim(token, "preferred_username");
        } catch (Exception e) {
            log.error("Error extracting preferred_username from token", e);
            return null;
        }
    }

    public String getFullNameFromToken(String token) {
        try {
            return (String) extractClaim(token, "name");
        } catch (Exception e) {
            log.error("Error extracting name from token", e);
            return null;
        }
    }

    public boolean isEmailVerifiedFromToken(String token) {
        try {
            Object emailVerified = extractClaim(token, "email_verified");
            if (emailVerified instanceof Boolean) {
                return (Boolean) emailVerified;
            }
            return false;
        } catch (Exception e) {
            log.error("Error extracting email_verified from token", e);
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    public List<String> getRolesFromResourceAccess(String token, String clientId) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);
            Object resourceAccess = claimsSet.getClaim("resource_access");
            if (resourceAccess instanceof Map) {
                Map<String, Object> resourceAccessMap = (Map<String, Object>) resourceAccess;
                Object clientAccess = resourceAccessMap.get(clientId);
                if (clientAccess instanceof Map) {
                    Map<String, Object> clientAccessMap = (Map<String, Object>) clientAccess;
                    Object roles = clientAccessMap.get("roles");
                    if (roles instanceof List) {
                        return new ArrayList<>((List<String>) roles);
                    }
                }
            }
            return Collections.emptyList();
        } catch (Exception e) {
            log.error("Error extracting roles from resource_access for client: {}", clientId, e);
            return Collections.emptyList();
        }
    }

    @SuppressWarnings("unchecked")
    public List<String> getRealmRolesFromToken(String token) {
        try {
            JWTClaimsSet claimsSet = validateToken(token);
            Object realmAccess = claimsSet.getClaim("realm_access");
            if (realmAccess instanceof Map) {
                Map<String, Object> realmAccessMap = (Map<String, Object>) realmAccess;
                Object roles = realmAccessMap.get("roles");
                if (roles instanceof List) {
                    return new ArrayList<>((List<String>) roles);
                }
            }
            return Collections.emptyList();
        } catch (Exception e) {
            log.error("Error extracting realm roles from token", e);
            return Collections.emptyList();
        }
    }

    public boolean hasRealmRole(String token, String roleName) {
        List<String> realmRoles = getRealmRolesFromToken(token);
        return realmRoles.contains(roleName);
    }

    public boolean hasResourceRole(String token, String clientId, String roleName) {
        List<String> resourceRoles = getRolesFromResourceAccess(token, clientId);
        return resourceRoles.contains(roleName);
    }

    /**
     * Validates token without signature verification (for development/testing)
     */
    public JWTClaimsSet validateTokenWithoutSignature(String token) {
        try {
            SignedJWT signedJWT = SignedJWT.parse(token);
            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();

            if (claimsSet.getExpirationTime() != null && claimsSet.getExpirationTime().before(new Date())) {
                throw new RuntimeException("JWT token has expired");
            }

            if (claimsSet.getIssueTime() != null && claimsSet.getIssueTime().after(new Date())) {
                throw new RuntimeException("JWT token not yet valid");
            }

            return claimsSet;

        } catch (Exception e) {
            log.error("Error parsing JWT token", e);
            throw new RuntimeException("Invalid JWT token format", e);
        }
    }

    public Object extractClaimWithoutValidation(String token, String claimName) {
        try {
            JWTClaimsSet claimsSet = validateTokenWithoutSignature(token);
            return claimsSet.getClaim(claimName);
        } catch (Exception e) {
            log.error("Error extracting claim {} from token without validation", claimName, e);
            return null;
        }
    }

    public boolean isTokenValid(String token) {
        try {
            validateToken(token);
            return true;
        } catch (Exception e) {
            log.debug("Token validation failed: {}", e.getMessage());
            return false;
        }
    }

    public boolean validateIssuer(String token, String expectedIssuer) {
        try {
            JWTClaimsSet claimsSet = getClaimsFromToken(token);
            if (claimsSet == null)
                return false;

            String issuer = claimsSet.getIssuer();
            return expectedIssuer.equals(issuer);
        } catch (Exception e) {
            log.error("Error validating issuer", e);
            return false;
        }
    }

    public boolean validateAudience(String token, String expectedAudience) {
        try {
            JWTClaimsSet claimsSet = getClaimsFromToken(token);
            if (claimsSet == null)
                return false;

            List<String> audiences = claimsSet.getAudience();
            return audiences != null && audiences.contains(expectedAudience);
        } catch (Exception e) {
            log.error("Error validating audience", e);
            return false;
        }
    }


    public String getAuthorizedPartyFromToken(String token) {
        try {
            return (String) extractClaim(token, "azp");
        } catch (Exception e) {
            log.error("Error extracting azp from token", e);
            return null;
        }
    }

    public String getSessionIdFromToken(String token) {
        try {
            return (String) extractClaim(token, "sid");
        } catch (Exception e) {
            log.error("Error extracting sid from token", e);
            return null;
        }
    }
}

File: ptm_task/src/kernel/utils/jwt_utils.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package utils

import (
	"errors"
	"slices"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/golibs-starter/golib/log"
	"github.com/serp/ptm-task/src/kernel/properties"
)

type JWTUtils struct {
	keycloakProps     *properties.KeycloakProperties
	keycloakJwksUtils *KeycloakJwksUtils
}

func NewJWTUtils(keycloakProps *properties.KeycloakProperties, keycloakJwksUtils *KeycloakJwksUtils) *JWTUtils {
	return &JWTUtils{
		keycloakProps:     keycloakProps,
		keycloakJwksUtils: keycloakJwksUtils,
	}
}

type Claims struct {
	UserID            int64                  `json:"uid"`
	Email             string                 `json:"email"`
	FullName          string                 `json:"name"`
	PreferredUsername string                 `json:"preferred_username"`
	EmailVerified     bool                   `json:"email_verified"`
	RealmAccess       map[string]interface{} `json:"realm_access"`
	ResourceAccess    map[string]interface{} `json:"resource_access"`
	AuthorizedParty   string                 `json:"azp"`
	SessionId         string                 `json:"sid"`
	jwt.RegisteredClaims
}

func (j *JWTUtils) ValidateToken(tokenString string) (*Claims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &Claims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, errors.New("unexpected signing method")
		}

		keyId, ok := token.Header["kid"].(string)
		if !ok {
			log.Warn("No key ID found in JWT header, skipping signature verification")
			return nil, errors.New("no key ID in JWT header")
		}

		publicKey, err := j.keycloakJwksUtils.GetPublicKey(keyId)
		if err != nil {
			log.Warn("Could not find public key for key ID: ", keyId, ", error: ", err)
			return nil, err
		}

		return publicKey, nil
	})

	if err != nil {
		log.Error("Failed to parse JWT token: ", err)
		return nil, err
	}

	if claims, ok := token.Claims.(*Claims); ok && token.Valid {
		if claims.ExpiresAt != nil && time.Now().After(claims.ExpiresAt.Time) {
			log.Error("JWT token is expired")
			return nil, errors.New("token is expired")
		}

		if j.keycloakProps.ExpectedIssuer != "" && claims.Issuer != j.keycloakProps.ExpectedIssuer {
			log.Error("JWT token issuer mismatch. Expected: ", j.keycloakProps.ExpectedIssuer, ", Actual: ", claims.Issuer)
			return nil, errors.New("token issuer mismatch")
		}

		if j.keycloakProps.ExpectedAudience != "" {
			audienceFound := slices.Contains(claims.Audience, j.keycloakProps.ExpectedAudience)
			if !audienceFound {
				log.Error("JWT token audience mismatch. Expected: ", j.keycloakProps.ExpectedAudience, ", Actual: ", claims.Audience)
				return nil, errors.New("token audience mismatch")
			}
		}

		return claims, nil
	}

	return nil, errors.New("invalid token")
}

func (j *JWTUtils) ExtractUserID(tokenString string) (int64, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return 0, err
	}
	return claims.UserID, nil
}

func (j *JWTUtils) ExtractEmail(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.Email, nil
}

func (j *JWTUtils) ExtractFullName(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.FullName, nil
}

func (j *JWTUtils) ExtractRoles(tokenString string) ([]string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return nil, err
	}

	var roles []string

	if claims.RealmAccess != nil {
		if realmRoles, ok := claims.RealmAccess["roles"].([]interface{}); ok {
			for _, role := range realmRoles {
				if roleStr, ok := role.(string); ok {
					roles = append(roles, roleStr)
				}
			}
		}
	}

	if claims.ResourceAccess != nil {
		for _, clientAccess := range claims.ResourceAccess {
			if clientMap, ok := clientAccess.(map[string]interface{}); ok {
				if clientRoles, ok := clientMap["roles"].([]interface{}); ok {
					for _, role := range clientRoles {
						if roleStr, ok := role.(string); ok {
							roles = append(roles, roleStr)
						}
					}
				}
			}
		}
	}

	unique := make(map[string]bool)
	var result []string
	for _, role := range roles {
		if !unique[role] {
			unique[role] = true
			result = append(result, role)
		}
	}

	return result, nil
}

func (j *JWTUtils) HasRole(tokenString string, roleName string) bool {
	roles, err := j.ExtractRoles(tokenString)
	if err != nil {
		return false
	}

	roleUpper := strings.ToUpper(roleName)
	for _, role := range roles {
		if strings.ToUpper(role) == roleUpper {
			return true
		}
	}
	return false
}

func (j *JWTUtils) IsAccessToken(tokenString string) bool {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		keyId, _ := token.Header["kid"].(string)
		return j.keycloakJwksUtils.GetPublicKey(keyId)
	})

	if err != nil {
		return false
	}

	if tokenType, ok := token.Header["typ"].(string); ok {
		if tokenType != "JWT" {
			return false
		}
	}

	if mapClaims, ok := token.Claims.(jwt.MapClaims); ok {
		if tokenType, ok := mapClaims["typ"].(string); ok {
			return tokenType == "Bearer"
		}
	}

	return true
}

func (j *JWTUtils) IsRefreshToken(tokenString string) bool {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		keyId, _ := token.Header["kid"].(string)
		return j.keycloakJwksUtils.GetPublicKey(keyId)
	})

	if err != nil {
		return false
	}

	if tokenType, ok := token.Header["typ"].(string); ok {
		if tokenType != "JWT" {
			return false
		}
	}

	if mapClaims, ok := token.Claims.(jwt.MapClaims); ok {
		if tokenType, ok := mapClaims["typ"].(string); ok {
			return tokenType == "Refresh"
		}
	}

	return false
}

func (j *JWTUtils) IsTokenExpired(tokenString string) bool {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, errors.New("unexpected signing method")
		}

		keyId, ok := token.Header["kid"].(string)
		if !ok {
			return nil, errors.New("no key ID in JWT header")
		}

		return j.keycloakJwksUtils.GetPublicKey(keyId)
	})

	if err != nil {
		return true
	}

	if claims, ok := token.Claims.(jwt.MapClaims); ok {
		if exp, ok := claims["exp"].(float64); ok {
			return time.Now().Unix() > int64(exp)
		}
	}

	return true
}

// Keycloak-specific methods

func (j *JWTUtils) GetSubjectFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.Subject, nil
}

func (j *JWTUtils) GetEmailFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.Email, nil
}

func (j *JWTUtils) GetPreferredUsernameFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.PreferredUsername, nil
}

func (j *JWTUtils) GetFullNameFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.FullName, nil
}

func (j *JWTUtils) IsEmailVerifiedFromToken(tokenString string) (bool, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return false, err
	}
	return claims.EmailVerified, nil
}

func (j *JWTUtils) GetRealmRolesFromToken(tokenString string) ([]string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return nil, err
	}

	var roles []string
	if claims.RealmAccess != nil {
		if realmRoles, ok := claims.RealmAccess["roles"].([]interface{}); ok {
			for _, role := range realmRoles {
				if roleStr, ok := role.(string); ok {
					roles = append(roles, roleStr)
				}
			}
		}
	}

	return roles, nil
}

func (j *JWTUtils) GetResourceRolesFromToken(tokenString string, clientId string) ([]string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return nil, err
	}

	var roles []string
	if claims.ResourceAccess != nil {
		if clientAccess, ok := claims.ResourceAccess[clientId].(map[string]interface{}); ok {
			if clientRoles, ok := clientAccess["roles"].([]interface{}); ok {
				for _, role := range clientRoles {
					if roleStr, ok := role.(string); ok {
						roles = append(roles, roleStr)
					}
				}
			}
		}
	}

	return roles, nil
}

func (j *JWTUtils) HasRealmRole(tokenString string, roleName string) bool {
	roles, err := j.GetRealmRolesFromToken(tokenString)
	if err != nil {
		return false
	}
	return slices.Contains(roles, roleName)
}

func (j *JWTUtils) HasResourceRole(tokenString string, clientId string, roleName string) bool {
	roles, err := j.GetResourceRolesFromToken(tokenString, clientId)
	if err != nil {
		return false
	}
	return slices.Contains(roles, roleName)
}

func (j *JWTUtils) GetAuthorizedPartyFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.AuthorizedParty, nil
}

func (j *JWTUtils) GetSessionIdFromToken(tokenString string) (string, error) {
	claims, err := j.ValidateToken(tokenString)
	if err != nil {
		return "", err
	}
	return claims.SessionId, nil
}
